File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxSum = -math.inf  # handle all-negative trees

        def helper(node):
            if not node:
                return 0

            # Compute max gain from left and right (ignore negative paths)
            left_gain = max(helper(node.left), 0)
            right_gain = max(helper(node.right), 0)

            # Path sum through this node (could be the best)
            current_sum = node.val + left_gain + right_gain
            self.maxSum = max(self.maxSum, current_sum)

            # Return best path *starting* from this node
            return node.val + max(left_gain, right_gain)

        helper(root)
        return self.maxSum

File: Data Structures & Algorithms/invert-a-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    
    public TreeNode invertTree(TreeNode root) {
        if(root == null){
            return null;
        }
        TreeNode newNode = new TreeNode(root.val, null, null);
        helper(root, newNode);
        return newNode;
    }

    public void helper(TreeNode root, TreeNode newNode){
        if (root.right!=null){
            newNode.left = new TreeNode(root.right.val);
            helper(root.right, newNode.left);
        }

        if(root.left!=null){
            newNode.right = new TreeNode(root.left.val);
            helper(root.left, newNode.right);
        }
    }
}
